src/credentials/FabricTable.cpp

```diff
@@ -41,145 +42,156 @@ CHIP_ERROR FabricInfo::SetFabricLabel(const CharSpan & fabricLabel)
     return CHIP_NO_ERROR;
 }
 
+namespace {
+constexpr TLV::Tag kVendorIdTag    = TLV::ContextTag(0);
+constexpr TLV::Tag kFabricLabelTag = TLV::ContextTag(1);
+} // anonymous namespace
+
 CHIP_ERROR FabricInfo::CommitToStorage(PersistentStorageDelegate * storage)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
+    DefaultStorageKeyAllocator keyAlloc;
 
-    char key[kKeySize];
-    ReturnErrorOnFailure(GenerateKey(mFabric, key, sizeof(key)));
+    VerifyOrReturnError(CanCastTo<uint16_t>(mRootCert.size()) && CanCastTo<uint16_t>(mICACert.size()) &&
+                            CanCastTo<uint16_t>(mNOCCert.size()),
+                        CHIP_ERROR_BUFFER_TOO_SMALL);
 
-    StorableFabricInfo * info = chip::Platform::New<StorableFabricInfo>();
-    ReturnErrorCodeIf(info == nullptr, CHIP_ERROR_NO_MEMORY);
+    ReturnErrorOnFailure(
+        storage->SyncSetKeyValue(keyAlloc.FabricRCAC(mFabric), mRootCert.data(), static_cast<uint16_t>(mRootCert.size())));
 
-    info->mFabricIndex = mFabric;
-    info->mVendorId    = Encoding::LittleEndian::HostSwap16(mVendorId);
+    // If we stop storing ICA certs when empty, update LoadFromStorage
+    // accordingly to check for CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND.
+    ReturnErrorOnFailure(
+        storage->SyncSetKeyValue(keyAlloc.FabricICAC(mFabric), mICACert.data(), static_cast<uint16_t>(mICACert.size())));
 
-    size_t stringLength = strnlen(mFabricLabel, kFabricLabelMaxLengthInBytes);
-    memcpy(info->mFabricLabel, mFabricLabel, stringLength);
-    info->mFabricLabel[stringLength] = '\0'; // Set null terminator
+    ReturnErrorOnFailure(
+        storage->SyncSetKeyValue(keyAlloc.FabricNOC(mFabric), mNOCCert.data(), static_cast<uint16_t>(mNOCCert.size())));
 
+    Crypto::P256SerializedKeypair serializedOpKey;
     if (mOperationalKey != nullptr)
     {
-        SuccessOrExit(err = mOperationalKey->Serialize(info->mOperationalKey));
+        ReturnErrorOnFailure(mOperationalKey->Serialize(serializedOpKey));
     }
     else
     {
+        // Could we just not store it instead?  What would deserialize need to
+        // do then?
```

Comment:
If you try to store this key, it for sure will not match NOC. It should be pre-checked that a key is available before attempting to store a FabricInfo

Reply:
I generally agree; I was just preserving the existing logic for now.  Filed https://github.com/project-chip/connectedhomeip/issues/16035 and will fix that as part of the changes to fabric indexing, which will affect what we are storing.